src/v/bytes/details/io_fragment.h

```diff
@@ -94,14 +97,271 @@ class io_fragment {
+    public:
+        using iterator_category = std::bidirectional_iterator_tag;
+        using container_type = typename std::
+          conditional_t<Const, const io_fragment_list, io_fragment_list>;
+        using value_type =
+          typename std::conditional_t<Const, const io_fragment, io_fragment>;
+        using difference_type = std::ptrdiff_t;
+        using pointer = value_type*;
+        using reference = value_type&;
+
+    private:
+        iter(container_type* container, pointer curr)
+          : _current(curr)
+#ifndef NDEBUG
+          , _my_generation(container->_generation)
+#endif
+          , _container(container) {
+        }
+
+    public:
+        iter() noexcept = default;
+        // Support conversion to const_iterator
+        // NOLINTNEXTLINE(hicpp-explicit-conversions)
+        operator iter<true, Forward>() const {
+            check_generation();
+            return {_container, _current};
+        }
+
+        reference operator*() const {
+            check_generation();
+            return *_current;
+        }
+        pointer operator->() const {
+            check_generation();
+            return _current;
+        }
+        iter& operator++() {
+            if constexpr (Forward) {
+                _current = _current->_next;
+            } else {
+                _current = _current->_prev;
+            }
+            return *this;
+        }
+        iter operator++(int) {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+        iter& operator--() {
+            if constexpr (Forward) {
+                if (_current == nullptr) {
+                    _current = _container->_tail;
+                } else {
+                    _current = _current->_prev;
+                }
+            } else {
+                if (_current == nullptr) {
+                    _current = _container->_head;
+                } else {
+                    _current = _current->_next;
+                }
+            }
+            return *this;
+        }
+        iter operator--(int) {
+            auto tmp = *this;
+            --*this;
+            return tmp;
+        }
+        bool operator==(const iter& o) const { return _current == o._current; }
+        bool operator!=(const iter& o) const { return _current != o._current; }
+
+    private:
+        friend class io_fragment_list;
+
+        inline void check_generation() const {
+#ifndef NDEBUG
+            vassert(
+              _container->_generation == _my_generation,
+              "Attempting to use an invalidated iterator. The corresponding "
+              "iobuf has been mutated since this iterator was constructed.");
+#endif
+        }
+
+        pointer _current = nullptr;
+#ifndef NDEBUG
+        size_t _my_generation = 0;
+#endif
+        container_type* _container = nullptr;
+    };
+
+public:
+    io_fragment_list() noexcept = default;
+    io_fragment_list(io_fragment_list&&) noexcept = default;
```

Comment:
* don't we need to reset _head/_tail on the moved-from to satisfy the assertion in the destructor?

* also i wonder if there is any code in the tree that assumes a moved-from iobuf is empty, which the default ctor wouldn't do right?

maybe i'm missing something...

Reply:
It's only safe because `iobuf` "fixes" it: https://github.com/redpanda-data/redpanda/blob/317787e74b17973f6d2a7a71681266f6b374b3ae/src/v/bytes/iobuf.h#L94

Reply:
oh of course, this is replacing `container_t`, right

Reply:
would it be preferable in any way to push that down into the list's move constructor? genuine question, I don't think I have an opinion. probably roughly (or exactly) the same code?

Reply:
I would prefer to have that here, local. The problem is I don't know how to do that for move assignment, because this is a non owning container, I can't do anything but assert that the container is empty before.

Good news we don't use move assignment, so I can just delete it and handle move ctor here

Reply:
Oh this was a great suggestion, it's driven the benchmarks down even further:

| test                    |  iterations |      median |         mad |         min |         max |      allocs |       tasks |        inst |
| -                       |           - |           - |           - |           - |           - |           - |           - |           - |
| iobuf.move_bench_small  |   520848000 |     1.469ns |     0.000ns |     1.468ns |     1.471ns |       0.000 |       0.000 |         7.2 |
| iobuf.move_bench_medium |     1694000 |     1.582ns |     0.002ns |     1.579ns |     1.586ns |       0.000 |       0.000 |         7.2 |
| iobuf.move_bench_large  |      528000 |     1.599ns |     0.009ns |     1.590ns |     1.614ns |       0.000 |       0.000 |         7.2 |

Reply:
See the diff here: https://github.com/redpanda-data/redpanda/compare/57d1129b6c1916b291964c2e3a278b17c059eea6..944e7725b38b2dcee2cd4c49b19d06c19ade5024

Reply:
Hahaha sweet 🎉